### ISSUE TITLE
Save binary data as base64 to CSV

### ISSUE THREAD

**Issue by rslemos** [exiftool/exiftool#11]
To save binary data as base64 with json one can use `-b` with `-j`. But `-b` and `-csv` are incompatible. Can't we make them compatible the way it is with json?

Alternatively, if there is some deep incompatibility I'm not aware of, perhaps there could be a way to remove all binary data from `-csv`, because when exporting all the data as of today, the generated csv is invalid.

**Comment by boardhead** [comment]
Can you give an example which generates an invalid CSV output?

**Comment by rslemos** [comment]
Running

`$ exiftool -n -csv -G:4:3:2:1:0 -SourceFile -Main:ExifTool:ExifTool:ExifTool:ExifToolVersion -Main:ExifTool:ExifTool:ExifTool:Warning -Main:Camera:IFD0:EXIF:Make -Main:Camera:IFD0:EXIF:Model -Main:Image:IFD0:EXIF:Software -Main:Image:ExifIFD:EXIF:UserComment -Main:Preview:IFD1:EXIF:ThumbnailImage ./20181201_154019_001.jpg`

Gives the following output:

```
SourceFile,Main:ExifTool:ExifTool:ExifTool:ExifToolVersion,Main:ExifTool:ExifTool:ExifTool:Warning,Main:Camera:IFD0:EXIF:Make,Main:Camera:IFD0:EXIF:Model,Main:Image:IFD0:EXIF:Software,Main:Image:ExifIFD:EXIF:UserComment,Main:Preview:IFD1:EXIF:ThumbnailImage
./20181201_154019_001.jpg,11.16,Invalid EXIF text encoding for UserComment,samsung,SM-A910F,A910FUBU1CRJ1,"

ALCSIIF5��(...)������������������������",(Binary data 22537 bytes)
```

With is a botched csv utf-8 encoded file, as `$ iconv -f utf-8 -t utf-8 >/dev/null` denounces:
```
iconv: illegal input sequence at position 379
```
(position 378 is `5` - after "ALCSIIF")


With `-json` (instead of `-csv`) it comes as
[{
  "SourceFile": "./20181201_154019_001.jpg",
  "Main:ExifTool:ExifTool:ExifTool:ExifToolVersion": 11.16,
  "Main:ExifTool:ExifTool:ExifTool:Warning": "Invalid EXIF text encoding for UserComment",
  "Main:Camera:IFD0:EXIF:Make": "samsung",
  "Main:Camera:IFD0:EXIF:Model": "SM-A910F",
  "Main:Image:IFD0:EXIF:Software": "A910FUBU1CRJ1",
  "Main:Image:ExifIFD:EXIF:UserComment": "\n\u0000\n\u0000\u0000\u0000ALCSIIF5??\u0007\u0000?\u0015\u0002\u0000\u0000\u0000\u0001\u0000??\u0001\u0000??\u0001\u0000eK(...)?????????",
  "Main:Preview:IFD1:EXIF:ThumbnailImage": "(Binary data 22537 bytes, use -b option to extract)"
}]

(at least, not a botched json file)

The offending field is `Main:Image:ExifIFD:EXIF:UserComment` (which is unreadably encoded in the image file, as warned by `Main:ExifTool:ExifTool:ExifTool:Warning`).

In the sample above, I truncated the middle of its contents (replaced with `(...)`) because: i) it would get unnecessarily big and ii) I don't know what information is being conveyed in that binary data.

Also one can note that both `-csv` and `-json` refuse to output `Main:Preview:IFD1:EXIF:ThumbnailImage` field contents. It is a safe bet, of course.

But since I'm using this csv output to save the (somewhat readable) original EXIF data for future comparison after curating the images, it would be nice to have all information correctly written to this database. Even the binary ones. Even the unreadable ones (G-d knows when they'll become readable by some new software).

To make binary data compatible with csv output, it must be encoded somehow. I see that `-json` encodes some (UTF-8 wise) illegal byte sequences as `\uDDDD` (kind of java syntax; what's the name of this encoding again?). `-csv` pastes the binary data uncurated, falling victim of any illegal utf-8 sequence that may arise (or worse, some rogue `,` or `"` or `\n` destroying the entire csv structure).

It would be nice to have either one of 3 alternatives (or all of them, perhaps):
1. binary data get encoded before going to csv (or json or xml) output: there are [a handful binary to text encodings](https://en.wikipedia.org/wiki/Binary-to-text_encoding), though base64 or that `\uDDDD` thing is enough;
2. binary data is replaced by that message `(Binary data n bytes....)`: not as useful, but at least not botched output;
3. command line option to remove offending fields selectively: the same as above.

Hope that helps.
If you need some more info, just go ask for it. If you need help implementing it, just give me directions and I'll be happy to send you a patch for your review and a merge request.

**Comment by rslemos** [comment]
Just found that `--Main:Image:ExifIFD:EXIF:UserComment` does exactly what option 3 above asks for.

Also `-b` switches from behavior 2 to 1 above (but it is compatible only with `-json`, not with `-csv`, which is what this issue originally asks for).

And as a side effect of my investigation, I found that exiftool doesn't recognize `Main:Image:ExifIFD:EXIF:UserComment` as a potentially binary value (though it gets treatment with `-json` but not with `-csv`; anyway, this treatment deviates from behavior found in field `Main:Preview:IFD1:EXIF:ThumbnailImage`). This one could be filed as another issue.

I'm still open to help you implement the `-b -json` compatibility.

**Comment by boardhead** [comment]
The JSON specification states that the encoding must be UTF-8, so ExifTool ensures that this is so, and uses JSON character escaping when necessary.

There is no specification for the encoding of CSV as far as I know, so in theory quoted binary data should be possible as long as the embedded quotes are escaped.  So I don't see the output you gave as being invalid.  If this is wrong, please point me to a specification.

I don't know what you mean by implementing the `-b -json` compatibility.  You mean in CSV?  If so, I don't see the point unless there is a CSV specification I am missing.

**Comment by boardhead** [comment]
Another consideration is that I really try to discourage people from using CVS because the format has some fairly serious drawbacks.  Not the least of which is that ExifTool must buffer output from all files in memory before it can be output as CVS, so it is very memory intensive if you are parsing a large number of files.

**Comment by rslemos** [comment]
> so in theory quoted binary data should be possible as long as the embedded quotes are escaped

I beg to differ.

CSV, being a text format, is not compatible with binary data. A kind of apples and oranges problem. To understand this it is important to have in mind that [characters](https://www.joelonsoftware.com/2003/10/08/the-absolute-minimum-every-software-developer-absolutely-positively-must-know-about-unicode-and-character-sets-no-excuses/) [are](https://www.oreilly.com/library/view/fluent-python/9781491946237/ch04.html) [not](https://docs.oracle.com/javase/tutorial/i18n/text/stream.html) [bytes](https://en.wikipedia.org/wiki/Character_encoding).

And the problem can be detected at the very heart of "escaping quotes embedded" in binary data (when pasting it to CSV text, inside double quotes).

Given the following binary data `0x00 0x63 0x00 0x22 0x00 0x2c 0x00 0x22 0x00 0xFF` how do you detect the quotes that are to be escaped? How do you interpret "quotes" inside this stream? Sequence `0x22 0x00` is a quote, by UCS-2LE; `0x00 0x22` is a quote, by UCS-2BE; byte `0x22` alone is a quote in most single byte charsets (many exceptions, for example, are some exotic encodings like EBCDIC family of charsets).

I propose `exiftool` to encode binary to text (base64, for instance), and then the whole CSV text (non binary fields, commas, quotation marks, line breaks and base64 encoded binary fields) to be encoded as bytes (the charset to be used can be chosen by command line switch, `-charset` for example).

**Comment by boardhead** [comment]
I managed to find a CSV specification:  https://tools.ietf.org/html/rfc4180

It does state that TEXTDATA should be %x23-2B / %x2D-7E

...which sort of precludes the possibility of anything other than plain ASCII.

However, it does mention the use of a "charset" MIME-type parameter, which is the only indication that charsets other than ASCII are a possibility, and sort of contradicts the specified restrictions on TEXTDATA.

Your example of a UCS-2BE output is irrelevant because ExifTool only outputs single-byte character encodings or UTF-8.  However, I get your point that software expecting text may have problems with embedded binary data in a CSV.  But why are you trying to embed binary in CVS in the first place?  Wouldn't it be preferable to just use JSON or XML instead?

**Comment by rslemos** [comment]
I'm using csv utilities to load the metadata into a relational database.

**Comment by rslemos** [comment]
I like RFC4180 because it is an attempt to bring some specification to a format that otherwise grew up organically.

But it doesn't clearly separate the "text" and the "octet" concerns. The entire format is text (as implied even by the chosen "text/csv" media type and implied RFC2046 section 4.1 which governs the "text/" subtree), but its grammar is specified in terms of bytes (`%x20-21`, `%x23-2B` and so on).

The RFC4180 itself has provisions for the `charset` parameter, so why specify TEXTDATA in terms of a byte range?

IMHO it is a safe and healthy interpretation to consider the CSV format as text (i.e. a stream/list of characters), and encode that text using whichever `charset` [chosen by the user among the ones] the application is capable of.

And given that it is text, no binary data can be embedded within, unless previously encoded by a binary-to-text encoding (which one is completely outside the scope of CSV "spec").

**Comment by boardhead** [comment]
OK.  You're starting to convince me.  But I need a set of clear rules to be able to implement this.  Something like this maybe?:

1. If ExifTool Charset is UTF8, then output a value as text if it is valid UTF-8, otherwise encode as Base64-encoded text with a leading string of "base64:".

2. If ExifTool Charset is other than UTF8, output a value as text if all characters are in the range 0x09,0x0a,0x0d,0x20-0x7e,0x80-0xfe, otherwise encode as Base64 as above.

**Comment by rslemos** [comment]
Hummm... I don't think that the decision to encode as base64 should be taken based on actual contents. 

I don't know the exact specifications of EXIF but I imagine the fields may have a datatype (or in case of UserComment perhaps another field describe how its contents is encoded). If this datatype is something like "binary" or "octet" or anything besides known datatypes which have a clear text representation (integers, floating points, rgb triplets, array of numbers and so on), then base64 encoding should kick in (basically ExifTool saying "I don't know what is this").

**Comment by boardhead** [comment]
That might work in an ideal world, but the real world of metadata is anything but ideal, and a tag that is supposed to hold text will often have other junk in it (and visa versa).

**Comment by rslemos** [comment]
I can only imagine what cameras in the wild are writing in EXIF fields...

Then your proposal seems fine to me. A safe bet.
Perhaps it deserves a warning "invalid data in field XXX; writing stream of bytes as base64".

**Comment by boardhead** [comment]
Adding this warning would be difficult because this check would need to be done at a lower level than the CSV formatting.

Also, maintaining backward compatibility is a concern, so what I propose is to leave the current ExifTool behaviour unchanged unless -charset or -b is combined with -csv.  If -charset is used, then the values are checked for validity and encoded as Base64 if necessary.  If -b is used then binary data is output, encoded as base64 if necessary.

**Comment by boardhead** [comment]
Resolved (hopefully) by ExifTool 11.29
